File: node/starter/template/src/complete_manual_aml_check.ts
import {type Client, NetworkService} from "@t-0/provider-sdk";
import {decimalToString} from "./lib";

export default async function completeManualAmlCheck(
    networkClient: Client<typeof NetworkService>,
    paymentId: bigint,
): Promise<void> {
  // Pay-Out Provider role — Step 2.6 (optional). If your payOut handler returned
  // manualAmlCheck instead of accepted (see ./service.ts), run your AML check
  // out-of-band and report the outcome with this call.
  const response = await networkClient.completeManualAmlCheck({
    paymentId,
    result: {case: 'approved', value: {}},
    // If your check failed, report a rejection instead:
    // result: {case: 'rejected', value: {reason: 'AML check failed'}},
  })

  switch (response.result.case) {
    case 'approved': {
      const approved = response.result.value;
      // Pay-in provider approved the updated quotes — proceed with the payout using
      // the updated amounts, then report the outcome via finalizePayout.
      const payOutAmount = approved.payOutAmount === undefined ? "not provided" : decimalToString(approved.payOutAmount);
      const settlementAmount = approved.settlementAmount === undefined ? "not provided" : decimalToString(approved.settlementAmount);
      console.log(`Manual AML check approved for payment ${paymentId}: quoteId=${approved.payOutQuoteId}, clientQuoteId=${approved.payOutClientQuoteId}, payOutAmount=${payOutAmount}, settlementAmount=${settlementAmount}`)
      break;
    }
    case 'rejected':
      // Pay-in provider rejected the updated quotes — do NOT proceed with the payout.
      console.log(`Updated quotes rejected for payment ${paymentId}, do not proceed with the payout`)
      break;
    default:
      console.error("unexpected result type")
  }
}

File: csharp/starter/template/Services/CompleteManualAmlCheck.cs
using T0.ProviderSdk.Api.Tzero.V1.Payment;

namespace {{PROJECT_NAME_PASCAL}}.Services;

// TODO: Step 2.6 (optional) Complete manual AML checks.
// Pay-Out Provider role: if your PayOut handler returned ManualAmlCheck instead of
// Accepted (see Services/PaymentHandler.cs), run your AML check out-of-band and
// report the outcome with this call.
public static class CompleteManualAmlCheck
{
    public static async Task CompleteAsync(NetworkService.NetworkServiceClient client, ulong paymentId)
    {
        try
        {
            var response = await client.CompleteManualAmlCheckAsync(new CompleteManualAmlCheckRequest
            {
                PaymentId = paymentId,
                Approved = new CompleteManualAmlCheckRequest.Types.Approved()
                // If your check failed, report a rejection instead:
                // Rejected = new CompleteManualAmlCheckRequest.Types.Rejected { Reason = "AML check failed" }
            });

            switch (response.ResultCase)
            {
                case CompleteManualAmlCheckResponse.ResultOneofCase.Approved:
                    // Pay-in provider approved the updated quotes — proceed with the payout using
                    // the updated amounts, then report the outcome via FinalizePayout.
                    Console.WriteLine($"Manual AML check approved for payment {paymentId}: payOutAmount={response.Approved.PayOutAmount}, settlementAmount={response.Approved.SettlementAmount}, quoteId={response.Approved.PayOutQuoteId}, clientQuoteId={response.Approved.PayOutClientQuoteId}");
                    break;
                case CompleteManualAmlCheckResponse.ResultOneofCase.Rejected:
                    // Pay-in provider rejected the updated quotes — do NOT proceed with the payout.
                    Console.WriteLine($"Updated quotes rejected for payment {paymentId}, do not proceed with the payout");
                    break;
            }
        }
        catch (Grpc.Core.RpcException ex)
        {
            Console.WriteLine($"Error completing manual AML check: {ex.Status.StatusCode} - {ex.Message}");
        }
    }
}
